File: javascript/src/ProgressPrinter.ts
import { Envelope } from '@cucumber/messages'
import { Query } from '@cucumber/query'

import { formatStatusCharacter } from './formatting'
import { SummaryPrinter } from './SummaryPrinter'
import { CUCUMBER_THEME } from './theme'
import type { ProgressOptions } from './types'

const DEFAULT_OPTIONS: Required<ProgressOptions> = {
  includeAttachments: true,
  summarise: false,
  theme: CUCUMBER_THEME,
}

/**
 * Prints test progress as single-character status indicators
 *
 * @remarks
 * Outputs a character for each test step as they complete, providing a compact
 * real-time view of test progress. Characters are styled according to their status
 * (e.g. `.` for passed, `F` for failed etc).
 */
export class ProgressPrinter {
  private readonly stream: NodeJS.WritableStream
  private readonly print: (content: string) => void
  private readonly options: Required<ProgressOptions>
  private readonly query: Query = new Query()

  /**
   * Creates a new ProgressPrinter instance
   *
   * @param params - Initialisation object
   * @param params.stream - The stream being written to, used for feature detection
   * @param params.options - Options for the output
   */
  constructor({
    stream = process.stdout,
    options = {},
  }: {
    stream?: NodeJS.WritableStream
    options?: ProgressOptions
  } = {}) {
    this.stream = stream
    this.print = (content) => stream.write(content)
    this.options = {
      ...DEFAULT_OPTIONS,
      ...options,
    }
  }

  /**
   * Processes a Cucumber message envelope and prints if appropriate
   *
   * @param message - The Cucumber message envelope to process
   */
  update(message: Envelope) {
    this.query.update(message)

    if (message.testStepFinished) {
      this.print(
        formatStatusCharacter(
          message.testStepFinished.testStepResult.status,
          this.options.theme,
          this.stream
        )
      )
    }

    if (message.testRunHookFinished) {
      this.print(
        formatStatusCharacter(
          message.testRunHookFinished.result.status,
          this.options.theme,
          this.stream
        )
      )
    }

    if (message.testRunFinished) {
      this.print('\n')
      if (this.options.summarise) {
        this.summarise()
      }
    }
  }

  private summarise() {
    SummaryPrinter.summarise(this.query, {
      stream: this.stream,
      options: this.options,
    })
  }
}

File: ruby/lib/cucumber/pretty_formatter/progress_printer.rb
# frozen_string_literal: true

module Cucumber
  module PrettyFormatter
    class ProgressPrinter < Printer
      DEFAULT_OPTIONS = {
        include_attachments: true,
        summarise: false,
        theme: CUCUMBER_THEME
      }.freeze

      def initialize(stream: $stdout, **options)
        super(stream: stream, **DEFAULT_OPTIONS.merge(options))
      end

      def update(envelope)
        query.update(envelope)
        write_test_step_finished(envelope)
        write_test_run_hook_finished(envelope)
        stream.write("\n") if envelope_value(envelope, :test_run_finished)
        nil
      end

      private

      def write_test_step_finished(envelope)
        test_step_finished = envelope_value(envelope, :test_step_finished)
        write_status(test_step_finished.test_step_result.status) if test_step_finished
      end

      def write_test_run_hook_finished(envelope)
        test_run_hook_finished = envelope_value(envelope, :test_run_hook_finished)
        write_status(test_run_hook_finished.result.status) if test_run_hook_finished
      end

      def write_status(status)
        stream.write(Formatting.format_status_character(status, options.fetch(:theme), stream))
      end

      def envelope_value(envelope, name)
        envelope.public_send(name) if envelope.respond_to?(name)
      end
    end
  end
end
